datasync/chaindatafetcher/kafka/consumer.go

```diff
@@ -101,20 +158,127 @@ func (c *Consumer) Cleanup(s sarama.ConsumerGroupSession) error {
+// case3. duplicated segment is ignored.
+// case4. new sparse segment shouldn't be given, so return an error.
+func insertSegment(segment *Segment, buffer [][]*Segment) ([][]*Segment, error) {
+	for idx, segments := range buffer {
+		length := len(segments)
+		if length > 0 && segments[0].key == segment.key {
+			// there is a missing segment which should not exist.
+			if segment.index > uint64(length) {
+				return buffer, errors.New(missingSegmentErrorMsg)
+			}
+
+			// the segment is already inserted to buffer.
+			if segment.index < uint64(length) {
+				logger.Warn("the message is duplicated", "segment", segment)
+				return buffer, nil
+			}
+
+			// insert the segment to the buffer.
+			buffer[idx] = append(segments, segment)
+			return buffer, nil
+		}
+	}
+
+	if segment.index == 0 {
+		// create a segment slice and append it.
+		buffer = append(buffer, []*Segment{segment})
+	} else {
+		// the segment may be already handled.
+		logger.Warn("the message may be inserted already. drop the segment", "segment", segment)
+	}
+	return buffer, nil
+}
+
+// handleBufferedMessages handles all consecutive complete messages in the buffer.
+func (c *Consumer) handleBufferedMessages(buffer [][]*Segment) ([][]*Segment, error) {
+	for len(buffer) > 0 {
+		// if any message exists in the buffer
+		oldestMsg, firstSegment, buffered := buffer[0], buffer[0][0], len(buffer[0])
+		if uint64(buffered) != firstSegment.total {
+			// not ready for assembling messages
+			return buffer, nil
+		}
+
+		// ready for assembling message
+		var msgBuffer []byte
+		for _, segment := range oldestMsg {
+			msgBuffer = append(msgBuffer, segment.value...)
+		}
+		msg := &sarama.ConsumerMessage{
+			Key:   []byte(firstSegment.key),
+			Value: msgBuffer,
+		}
+
+		f, ok := c.handlers[firstSegment.orig.Topic]
+		if !ok {
+			return buffer, fmt.Errorf("%v: %v", noHandlerErrorMsg, msg.Topic)
+		}
+
+		if err := f(msg); err != nil {
+			return buffer, err
+		}
+
+		buffer = buffer[1:]
+	}
+
+	return buffer, nil
+}
+
+// updateOffset updates offset after handling messages.
+// The offset should be marked for the oldest message (which is not read) in the given buffer.
+// If there is no segment in the buffer, the last consumed message offset should be marked.
+func (c *Consumer) updateOffset(buffer [][]*Segment, lastMsg *sarama.ConsumerMessage, session ConsumerGroupSession) error {
+	if len(buffer) > 0 {
+		if len(buffer[0]) <= 0 {
+			return errors.New(emptySegmentErrorMsg)
+		}
+
+		oldestMsg := buffer[0][0].orig
+		session.MarkOffset(oldestMsg.Topic, oldestMsg.Partition, oldestMsg.Offset, "")
+	} else {
+		session.MarkMessage(lastMsg, "")
+	}
+
+	return nil
+}
+
 // ConsumeClaim must start a consumer loop of ConsumerGroupClaim's Messages().
 // Once the Messages() channel is closed, the Handler must finish its processing
 // loop and exit.
 func (c *Consumer) ConsumeClaim(cgs sarama.ConsumerGroupSession, cgc sarama.ConsumerGroupClaim) error {
```

Comment:
Any plans to create unit tests for `ConsumeClaim`?

Reply:
The `ConsumeClaim` is called in `Consume` in the `Subscribe` function. I will add more PubSub testcases.